Mathlib/GroupTheory/SpecificGroups/Coxeter.lean

```diff
@@ -0,0 +1,377 @@
+  presentation corresponding to a Coxeter matrix, i.e. `Matrix.CoxeterGroup B M`.
+* `IsCoxeterGroup` : A group is a Coxeter group if it is registered in a Coxeter System.
+
+## References
+
+* https://en.wikipedia.org/wiki/Coxeter_group
+
+* [N. Bourbaki, *Lie Groups and Lie Algebras, Chapters 4--6*](bourbaki1968) chapter IV
+  pages 13--15
+
+## Tags
+
+coxeter system, coxeter group
+-/
+
+noncomputable section
+
+variable {B : Type*} [DecidableEq B]
+
+variable (M : Matrix B B ℕ∞)
+
+variable (B)
+
+/-- A matrix `IsCoxeter` if it is a symmetric matrix with ones on the diagonal
+and off-diagonal elements are greater than or equal to two. -/
+class Matrix.IsCoxeter (M : Matrix B B ℕ∞) : Prop where
+  symmetric : M.IsSymm
+  diagonal : ∀ i j : B, i = j → M i j = 1
+  off_diagonal : ∀ i j : B, i ≠ j → 2 ≤ M i j
+
+namespace CoxeterGroup
+
+variable {B}
+
+namespace Relations
+
+/-- The relation terms corresponding to a Coxeter matrix. -/
+def ofMatrix : B × B → FreeGroup B :=
+  Function.uncurry fun i j =>
+    (FreeGroup.of i * FreeGroup.of j) ^ orderOf (M i j)
+
+/-- The relations corresponding to a Coxeter matrix. -/
+def toSet : Set (FreeGroup B) :=
+  Set.range <| ofMatrix M
+
+end Relations
+
+end CoxeterGroup
+
+/-- The group presentation corresponding to a Coxeter matrix.
+
+Note that it is defined for any matrix of natural numbers. Its value for non-Coxeter
+matrices should be regarded as junk. `IsCoxeterGroup` checks that the matrix `IsCoxeter`. -/
+abbrev Matrix.CoxeterGroup := PresentedGroup <| CoxeterGroup.Relations.toSet M
+
+/-- A Coxeter System `CoxeterSystem W` is a structure recording the isomorphism between
+a group `W` and the group presentation corresponding to a Coxeter matrix. Equivalently, this can
+be seen as a list of generators of `W` parameterized by the underlying type of `M`, which
+satisfy the relations of the Coxeter matrix `M`. -/
+structure CoxeterSystem (W : Type*) [Group W] where
+  /-- `CoxeterSystem.ofRepr` constructs a Coxeter system given an equivalence with the group
+  presentation corresponding to a Coxeter matrix `M`. -/
+  ofRepr ::
+    /-- `repr` is the isomorphism between the group `W` and the group presentation
+    corresponding to a Coxeter matrix `M`. -/
+    repr : Matrix.CoxeterGroup B M ≃* W
+
+/-- A group is a Coxeter group if it is registered in a Coxeter System. -/
+class IsCoxeterGroup (W : Type*) [Group W] : Prop where
+  nonempty_system : ∃ (M : Matrix B B ℕ∞), M.IsCoxeter ∧ Nonempty (CoxeterSystem B M W)
+
+namespace CoxeterMatrix
+
+variable {m n : ℕ}
+
+/-- Auxilary definition for `Bₙ`. -/
+def AₙAux : (Fin n) → (Fin n) → ℕ∞ :=
+  fun i j =>
+    if i == j then 1
+      else (if i == n - 1 ∨ j == n - 1 then 2 else 3)
+
+/-- The Coxeter matrix of family Aₙ.
+
+The corresponding Coxeter-Dynkin diagram is:
+```
+    o --- o --- o ⬝ ⬝ ⬝ ⬝ o --- o
+```
+-/
+def Aₙ : Matrix (Fin n) (Fin n) ℕ∞ :=
+  Matrix.of AₙAux
```

Comment:
There is no need for auxiliary definitions here. I am not completely sure what is the mathlib practice for the definition of matrices, but if you do `abbrev Aₙ : Matrix (Fin n) (Fin n) ℕ := ...`, then the proof of `IsCoxeter` will be automatic. I think that `Aₙ` (better name `A`) should take `n` as an explicit parameter rather than an implicit one.